File: internal/gitkit/ts/bin/pr-events.ts
// pr-events.ts — the self-terminating PR watcher for `prm`'s native Monitor.
// Run (under Monitor, persistent): node pr-events.ts <pr> [--every-seconds N]
//   → polls the PR and prints ONE event line per state delta, then EXITS when the
//     PR is merged/closed (the watch self-terminates — nothing to reap).
//
// Event vocabulary (one per line; the prm loop maps each to an action):
//   comment <id>                       new review-thread OR conversation comment
//   ci <prev>-><curr>                  CI rollup flip
//   review <STATE> by <login>          reviewer verdict change (APPROVED / CHANGES_REQUESTED / DISMISSED)
//   push <sha7>                        the PR head moved (incl. force-push)
//   mergeable <prev>-><curr>           MERGEABLE <-> CONFLICTING flips (UNKNOWN churn suppressed)
//   draft / ready                      draft state flips
//   merged / closed                    terminal — the watcher exits
//
// Durability: every `gh` call is timeout-bounded (a hung network call must not
// wedge the watch), startup retries, and poll failures back off exponentially
// (rate-limit-looking failures jump straight to the max backoff).
// Zero npm deps: shells out to `gh`, Node stdlib only. Erasable TS only.

// reviews: latest review STATE per reviewer login (APPROVED / CHANGES_REQUESTED /
// DISMISSED). commentIds covers BOTH unresolved review-thread comments and
// conversation (issue) comments — distinct databaseId spaces, one event stream.
// All new fields are optional for backward compatibility with older snapshots.
export type Snapshot = {
  state: string;
  ci: string;
  commentIds: number[];
  reviews?: Record<string, string>;
  headSha?: string;
  mergeable?: string; // MERGEABLE | CONFLICTING (UNKNOWN is carried over, never stored)
  isDraft?: boolean;
};
export type Events = { events: string[]; done: boolean };

function terminalEvent(state: string): string {
  return state.toLowerCase(); // "merged" | "closed"
}

export function computeEvents(prev: Snapshot | null, curr: Snapshot): Events {
  const events: string[] = [];
  const terminal = curr.state.toUpperCase() !== "OPEN";

  if (prev === null) {
    // Baseline poll: the main session already handled the existing backlog, so we emit
    // nothing for pre-existing comments — only fire if the PR is ALREADY terminal.
    if (terminal) return { events: [terminalEvent(curr.state)], done: true };
    return { events: [], done: false };
  }

  const seen = new Set(prev.commentIds);
  for (const id of curr.commentIds) if (!seen.has(id)) events.push(`comment ${id}`);
  if (prev.ci !== curr.ci) events.push(`ci ${prev.ci}->${curr.ci}`);
  // Review-state transitions: an APPROVED review with no inline comments used to
  // produce NO event (no new thread, no CI flip) and the watch stayed silent on
  // the exact "ready to merge" signal the loop exists for.
  const prevReviews = prev.reviews ?? {};
  for (const [login, state] of Object.entries(curr.reviews ?? {})) {
    if (prevReviews[login] !== state) events.push(`review ${state} by ${login}`);
  }
  // Head movement (a teammate/bot pushed, or a force-push rewrote the branch).
  if (prev.headSha !== undefined && curr.headSha !== undefined && prev.headSha !== curr.headSha) {
    events.push(`push ${curr.headSha.slice(0, 7)}`);
  }
  // Mergeability: only real MERGEABLE <-> CONFLICTING flips. GitHub reports
  // UNKNOWN while recomputing after every push — fetchSnapshot carries the
  // previous value forward instead of storing UNKNOWN, so no churn here.
  if (
    prev.mergeable !== undefined &&
    curr.mergeable !== undefined &&
    prev.mergeable !== curr.mergeable
  ) {
    events.push(`mergeable ${prev.mergeable}->${curr.mergeable}`);
  }
  if (prev.isDraft !== undefined && curr.isDraft !== undefined && prev.isDraft !== curr.isDraft) {
    events.push(curr.isDraft ? "draft" : "ready");
  }
  if (terminal) {
    events.push(terminalEvent(curr.state));
    return { events, done: true };
  }
  return { events, done: false };
}

import { execFileSync } from "node:child_process";

import { repoRoot } from "./repo-root.ts";

// Timeout-bounded: a hung gh/network call must never wedge the watch — the
// caller treats the throw as a failed poll and backs off.
function sh(cmd: string, args: string[]): string {
  return execFileSync(cmd, args, {
    encoding: "utf8",
    maxBuffer: 32 * 1024 * 1024,
    timeout: 120_000,
    cwd: repoRoot(),
  });
}

const QUERY = `
query($owner:String!, $repo:String!, $pr:Int!) {
  repository(owner:$owner, name:$repo) {
    pullRequest(number:$pr) {
      state
      commits(last: 1) { nodes { commit { statusCheckRollup { state } } } }
      reviewThreads(first: 100) {
        nodes { isResolved comments(first: 1) { nodes { databaseId } } }
      }
      latestReviews(first: 50) { nodes { state author { login } } }
      headRefOid
      mergeable
      isDraft
      comments(last: 50) { nodes { databaseId } }
    }
  }
}`;

function fetchSnapshot(owner: string, repo: string, pr: number): Snapshot {
  const out = sh("gh", ["api", "graphql", "-f", `query=${QUERY}`, "-f", `owner=${owner}`, "-f", `repo=${repo}`, "-F", `pr=${pr}`]);
  const d = JSON.parse(out).data.repository.pullRequest;
  const ci: string = d.commits?.nodes?.[0]?.commit?.statusCheckRollup?.state ?? "NONE";
  const commentIds: number[] = [];
  for (const t of d.reviewThreads?.nodes ?? []) {
    if (t.isResolved) continue;
    const id = t.comments?.nodes?.[0]?.databaseId;
    if (typeof id === "number") commentIds.push(id);
  }
  for (const c of d.comments?.nodes ?? []) {
    if (typeof c.databaseId === "number") commentIds.push(c.databaseId);
  }
  const reviews: Record<string, string> = {};
  for (const r of d.latestReviews?.nodes ?? []) {
    const login = r.author?.login;
    // COMMENTED reviews are already surfaced through their comment ids.
    if (typeof login === "string" && typeof r.state === "string" && r.state !== "COMMENTED") {
      reviews[login] = r.state;
    }
  }
  // UNKNOWN = GitHub is recomputing (fires after every push) — carry the last
  // known value forward so computeEvents only sees real flips.
  const mergeable: string =
    d.mergeable === "UNKNOWN" ? (prevMergeable ?? "MERGEABLE") : (d.mergeable ?? "MERGEABLE");
  prevMergeable = mergeable;
  return {
    state: d.state,
    ci,
    commentIds,
    reviews,
    headSha: typeof d.headRefOid === "string" ? d.headRefOid : undefined,
    mergeable,
    isDraft: d.isDraft === true,
  };
}
let prevMergeable: string | undefined;

/**
 * `--repo` accepts BOTH forms every other lib/git script uses:
 *  - an absolute (or ./-relative) repo path → the anchor for repoRoot(), which
 *    reads the same flag from argv; owner/name then comes from the startup
 *    `gh repo view` lookup run in that anchored dir → nameWithOwner stays "".
 *  - `owner/name` → pins the API target directly, but MUST be stripped from
 *    process.argv before the first repoRoot() call (repoRoot throws on a
 *    non-directory --repo value, which would fail every poll silently).
 * Anything else throws loudly at startup.
 */
export function parseRepoFlag(value: string): { nameWithOwner: string; stripFromArgv: boolean } {
  if (value === "" || value.startsWith("/") || value.startsWith(".")) {
    return { nameWithOwner: "", stripFromArgv: false };
  }
  if (/^[^/\s]+\/[^/\s]+$/.test(value)) return { nameWithOwner: value, stripFromArgv: true };
  throw new Error(`--repo expects owner/name or an absolute repo path, got: ${value}`);
}

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  const pr = Number(argv.find((a) => !a.startsWith("--")));
  if (!Number.isInteger(pr) || pr <= 0) throw new Error("usage: pr-events.ts <pr> [--repo <abs repo path>|owner/name] [--every-seconds N]");
  const everyIdx = argv.indexOf("--every-seconds");
  const everyMs = Math.max(30, everyIdx >= 0 ? Number(argv[everyIdx + 1]) : 30) * 1000;

  // --repo owner/name pins the target repo explicitly. Without it the repo is
  // derived from the cwd — which silently watched the WRONG repo's PR number
  // when a cross-repo watch was started from another project's directory
  // (burned live 2026-07-24: watching FE#340 from the BE repo polled BE#340,
  // an old closed PR, and the watch self-terminated with a bogus `closed`).
  const repoIdx = argv.indexOf("--repo");
  const parsed = parseRepoFlag(repoIdx >= 0 ? String(argv[repoIdx + 1] ?? "") : "");
  let nameWithOwner = parsed.nameWithOwner;
  if (parsed.stripFromArgv) {
    // owner/name form pins the API target directly — but repoRoot() also reads
    // --repo and throws on a non-directory value, which would fail EVERY poll
    // (silently: each throw reads as a transient failure and backs off forever).
    // Strip the pair from process.argv so repoRoot falls back to ambient cwd.
    const i = process.argv.indexOf("--repo");
    if (i !== -1) process.argv.splice(i, 2);
  }

  // Startup is retried too — a transient gh hiccup at launch must not kill a
  // watch that was meant to run for hours.
  for (let attempt = 1; nameWithOwner === ""; attempt++) {
    try {
      nameWithOwner = JSON.parse(sh("gh", ["repo", "view", "--json", "nameWithOwner"])).nameWithOwner;
      break;
    } catch (e) {
      if (attempt >= 5) throw e;
      process.stderr.write(`warn: startup repo lookup failed (attempt ${attempt}/5); retrying\n`);
      await sleep(Math.min(attempt * 15_000, 60_000));
    }
  }
  const [owner, repo] = nameWithOwner.split("/");

  let prev: Snapshot | null = null;
  let failures = 0;
  for (;;) {
    let curr: Snapshot;
    try {
      curr = fetchSnapshot(owner, repo, pr);
      failures = 0;
    } catch (e) {
      // One failed poll shouldn't kill the watch (transient gh/network) — log to
      // stderr (NOT the event stream) and back off exponentially; rate-limit-
      // looking failures jump straight to the max backoff.
      failures++;
      const msg = (e as Error).message ?? "";
      const rateLimited = /rate limit|429|403/i.test(msg);
      const factor = rateLimited ? 8 : Math.min(2 ** Math.min(failures, 3), 8);
      process.stderr.write(`warn: poll failed (${msg}); backoff x${factor}\n`);
      await sleep(everyMs * factor);
      continue;
    }
    const { events, done } = computeEvents(prev, curr);
    for (const e of events) process.stdout.write(e + "\n");
    prev = curr;
    if (done) break;
    await sleep(everyMs);
  }
}

if (import.meta.main) {
  main().catch((e) => {
    process.stderr.write(`error: ${(e as Error).message}\n`);
    process.exit(1);
  });
}

File: internal/gitkit/prevents.go
package gitkit

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"time"
)

// pr-events — the self-terminating PR watcher behind prm's native Monitor:
// `pr-events <pr> [--every-seconds N] [--repo <abs path>|owner/name]` polls
// the PR and prints ONE event line per state delta, then EXITS when the PR is
// merged or closed. Event vocabulary (the prm loop maps each to an action):
//
//	comment <id>              new review-thread OR conversation comment
//	ci <prev>-><curr>         CI rollup flip
//	review <STATE> by <login> reviewer verdict change
//	push <sha7>               the PR head moved (incl. force-push)
//	mergeable <prev>-><curr>  MERGEABLE <-> CONFLICTING (UNKNOWN suppressed)
//	draft / ready             draft state flips
//	merged / closed           terminal — the watcher exits
//
// Durability: every gh call is timeout-bounded, startup retries, and poll
// failures back off exponentially (rate-limit-looking ones jump to the max).

// Snapshot is one poll's state. CommentIDs covers unresolved review-thread
// comments AND conversation comments — distinct id spaces, one stream. The
// optional fields are nil in older snapshots and then never fire.
type Snapshot struct {
	State      string
	CI         string
	CommentIDs []int
	Reviews    *reviewStates // latest non-COMMENTED review state per login
	HeadSha    *string
	Mergeable  *string // MERGEABLE | CONFLICTING (UNKNOWN is carried over, never stored)
	IsDraft    *bool
}

// reviewStates is a login → state map iterated in JavaScript object order:
// integer-like keys ascending, then the rest in insertion order.
type reviewStates struct {
	keys  []string
	state map[string]string
}

var arrayIndex = regexp.MustCompile(`^(0|[1-9]\d*)$`)

func (r *reviewStates) set(login, state string) {
	if r.state == nil {
		r.state = map[string]string{}
	}
	if _, ok := r.state[login]; !ok {
		r.keys = append(r.keys, login)
	}
	r.state[login] = state
}

func (r *reviewStates) ordered() []string {
	if r == nil {
		return nil
	}
	isIndex := func(k string) bool {
		n, err := strconv.ParseUint(k, 10, 64)
		return arrayIndex.MatchString(k) && err == nil && n < math.MaxUint32
	}
	var indices, rest []string
	for _, k := range r.keys {
		if isIndex(k) {
			indices = append(indices, k)
		} else {
			rest = append(rest, k)
		}
	}
	slices.SortFunc(indices, func(a, b string) int {
		x, _ := strconv.ParseUint(a, 10, 64)
		y, _ := strconv.ParseUint(b, 10, 64)
		return int(x) - int(y)
	})
	return append(indices, rest...)
}

func (r *reviewStates) get(login string) (string, bool) {
	if r == nil {
		return "", false
	}
	s, ok := r.state[login]
	return s, ok
}

type threadNode struct {
	IsResolved bool `json:"isResolved"`
	Comments   *struct {
		Nodes []struct {
			DatabaseID *int   `json:"databaseId"`
			Author     *login `json:"author"`
		} `json:"nodes"`
	} `json:"comments"`
}

// openThreadCommentIDs is every comment in an unresolved thread, not just
// its root: a reviewer replying inside an open thread changes nothing else in
// the snapshot. The watcher's own replies are round output, not new work.
func openThreadCommentIDs(threads []threadNode, self string) []int {
	ids := []int{}
	for _, t := range threads {
		if t.IsResolved || t.Comments == nil {
			continue
		}
		for _, c := range t.Comments.Nodes {
			if c.DatabaseID != nil && (c.Author == nil || c.Author.Login != self) {
				ids = append(ids, *c.DatabaseID)
			}
		}
	}
	return ids
}

// computeEvents diffs two polls. The baseline poll (prev nil) is silent —
// the session already handled the backlog — unless the PR is terminal.
func computeEvents(prev *Snapshot, curr Snapshot) ([]string, bool) {
	events := []string{}
	terminal := strings.ToUpper(curr.State) != "OPEN"
	if prev == nil {
		if terminal {
			return []string{strings.ToLower(curr.State)}, true
		}
		return events, false
	}
	for _, id := range curr.CommentIDs {
		if !slices.Contains(prev.CommentIDs, id) {
			events = append(events, fmt.Sprintf("comment %d", id))
		}
	}
	if prev.CI != curr.CI {
		events = append(events, fmt.Sprintf("ci %s->%s", prev.CI, curr.CI))
	}
	// An APPROVED review with no inline comments used to produce NO event and
	// the watch stayed silent on the exact "ready to merge" signal.
	for _, who := range curr.Reviews.ordered() {
		state, _ := curr.Reviews.get(who)
		if was, ok := prev.Reviews.get(who); !ok || was != state {
			events = append(events, fmt.Sprintf("review %s by %s", state, who))
		}
	}
	if prev.HeadSha != nil && curr.HeadSha != nil && *prev.HeadSha != *curr.HeadSha {
		events = append(events, "push "+string(jsSlice(*curr.HeadSha, 7).s))
	}
	// Only real MERGEABLE <-> CONFLICTING flips: fetchSnapshot carries the
	// previous value over GitHub's post-push UNKNOWN.
	if prev.Mergeable != nil && curr.Mergeable != nil && *prev.Mergeable != *curr.Mergeable {
		events = append(events, fmt.Sprintf("mergeable %s->%s", *prev.Mergeable, *curr.Mergeable))
	}
	if prev.IsDraft != nil && curr.IsDraft != nil && *prev.IsDraft != *curr.IsDraft {
		if *curr.IsDraft {
			events = append(events, "draft")
		} else {
			events = append(events, "ready")
		}
	}
	if terminal {
		return append(events, strings.ToLower(curr.State)), true
	}
	return events, false
}

const eventsQuery = `
query($owner:String!, $repo:String!, $pr:Int!) {
  viewer { login }
  repository(owner:$owner, name:$repo) {
    pullRequest(number:$pr) {
      state
      commits(last: 1) { nodes { commit { statusCheckRollup { state } } } }
      reviewThreads(first: 100) {
        nodes { isResolved comments(last: 50) { nodes { databaseId author { login } } } }
      }
      latestReviews(first: 50) { nodes { state author { login } } }
      headRefOid
      mergeable
      isDraft
      comments(last: 50) { nodes { databaseId } }
    }
  }
}`

// shapeSnapshot turns the GraphQL answer into a Snapshot. prevMergeable is
// the last stored mergeability: UNKNOWN (GitHub recomputing after every
// push) carries it forward so computeEvents only sees real flips.
func shapeSnapshot(out []byte, prevMergeable *string) (Snapshot, error) {
	var resp struct {
		Data *struct {
			Viewer *login `json:"viewer"`
			Repo   *struct {
				PR *struct {
					State   string `json:"state"`
					Commits *struct {
						Nodes []struct {
							Commit *struct {
								StatusCheckRollup *struct {
									State *string `json:"state"`
								} `json:"statusCheckRollup"`
							} `json:"commit"`
						} `json:"nodes"`
					} `json:"commits"`
					ReviewThreads *struct {
						Nodes []threadNode `json:"nodes"`
					} `json:"reviewThreads"`
					LatestReviews *struct {
						Nodes []struct {
							State  any `json:"state"`
							Author *struct {
								Login any `json:"login"`
							} `json:"author"`
						} `json:"nodes"`
					} `json:"latestReviews"`
					HeadRefOid any     `json:"headRefOid"`
					Mergeable  *string `json:"mergeable"`
					IsDraft    any     `json:"isDraft"`
					Comments   *struct {
						Nodes []struct {
							DatabaseID *int `json:"databaseId"`
						} `json:"nodes"`
					} `json:"comments"`
				} `json:"pullRequest"`
			} `json:"repository"`
		} `json:"data"`
	}
	if err := json.Unmarshal(out, &resp); err != nil {
		return Snapshot{}, err
	}
	if resp.Data == nil || resp.Data.Repo == nil || resp.Data.Repo.PR == nil {
		return Snapshot{}, errors.New("graphql response carries no pullRequest")
	}
	d := resp.Data.Repo.PR
	s := Snapshot{State: d.State, CI: "NONE", Reviews: &reviewStates{}}
	if d.Commits != nil && len(d.Commits.Nodes) > 0 {
		if c := d.Commits.Nodes[0].Commit; c != nil && c.StatusCheckRollup != nil && c.StatusCheckRollup.State != nil {
			s.CI = *c.StatusCheckRollup.State
		}
	}
	self := ""
	if resp.Data.Viewer != nil {
		self = resp.Data.Viewer.Login
	}
	s.CommentIDs = []int{}
	if d.ReviewThreads != nil {
		s.CommentIDs = openThreadCommentIDs(d.ReviewThreads.Nodes, self)
	}
	if d.Comments != nil {
		for _, c := range d.Comments.Nodes {
			if c.DatabaseID != nil {
				s.CommentIDs = append(s.CommentIDs, *c.DatabaseID)
			}
		}
	}
	if d.LatestReviews != nil {
		for _, r := range d.LatestReviews.Nodes {
			// COMMENTED reviews already surface through their comment ids.
			state, stateOK := r.State.(string)
			if r.Author == nil || !stateOK || state == "COMMENTED" {
				continue
			}
			if who, ok := r.Author.Login.(string); ok {
				s.Reviews.set(who, state)
			}
		}
	}
	mergeable := "MERGEABLE"
	switch {
	case d.Mergeable != nil && *d.Mergeable == "UNKNOWN":
		if prevMergeable != nil {
			mergeable = *prevMergeable
		}
	case d.Mergeable != nil:
		mergeable = *d.Mergeable
	}
	s.Mergeable = &mergeable
	if sha, ok := d.HeadRefOid.(string); ok {
		s.HeadSha = &sha
	}
	draft := d.IsDraft == true
	s.IsDraft = &draft
	return s, nil
}

var ownerName = regexp.MustCompile(`^[^/\s]+/[^/\s]+$`)

// parseRepoFlag accepts BOTH --repo forms: a path (absolute, ./-relative or
// empty) is the anchor for repoRoot; `owner/name` pins the API target and
// must be stripped from the argv repoRoot reads — it would reject a
// non-directory and starve every poll. Anything else fails at startup.
func parseRepoFlag(value string) (nameWithOwner string, strip bool, err error) {
	if value == "" || strings.HasPrefix(value, "/") || strings.HasPrefix(value, ".") {
		return "", false, nil
	}
	if ownerName.MatchString(value) {
		return value, true, nil
	}
	return "", false, fmt.Errorf("--repo expects owner/name or an absolute repo path, got: %s", value)
}

// pollInterval is `--every-seconds N` with a 30 s floor. A value that is not
// a number, or past what a timer can hold, is the default — Node's NaN or
// overflowing setTimeout would poll every millisecond.
func pollInterval(args []string) time.Duration {
	seconds := 30.0
	if i := slices.Index(args, "--every-seconds"); i >= 0 && i+1 < len(args) {
		if n, ok := jsNumber(args[i+1]); ok && n > seconds && n*1000 <= math.MaxInt32 {
			seconds = n
		}
	}
	return time.Duration(seconds * float64(time.Second))
}

var rateLimitish = regexp.MustCompile(`(?i)rate limit|429|403`)

// sleep is the watcher's clock; tests replace it.
var sleep = time.Sleep

func runPREvents(args []string, stdout, stderr io.Writer) int {
	prArg := ""
	for _, a := range args {
		if !strings.HasPrefix(a, "--") {
			prArg = a
			break
		}
	}
	pr, ok := positiveInt(prArg)
	if !ok || prArg == "" {
		return fail(stderr, errors.New("usage: pr-events.ts <pr> [--repo <abs repo path>|owner/name] [--every-seconds N]"))
	}
	every := pollInterval(args)

	// --repo owner/name pins the target. Without it the repo is derived from
	// the anchor — a cross-repo watch from another project's directory once
	// polled the WRONG repo's PR #340 and exited on a bogus `closed`.
	repoValue := ""
	if i := slices.Index(args, "--repo"); i >= 0 && i+1 < len(args) {
		repoValue = args[i+1]
	}
	nameWithOwner, strip, err := parseRepoFlag(repoValue)
	if err != nil {
		return fail(stderr, err)
	}
	anchorArgs := args
	if strip {
		i := slices.Index(args, "--repo")
		anchorArgs = slices.Delete(slices.Clone(args), i, min(i+2, len(args)))
	}
	// The anchor resolves lazily and sticks once found, as every call would
	// otherwise re-resolve it; a failure is retried with the call.
	root := ""
	gh := func(a ...string) (string, error) {
		if root == "" {
			r, err := repoRoot(anchorArgs)
			if err != nil {
				return "", err
			}
			root = r
		}
		return execFile(execOpts{dir: root, echo: stderr, timeout: 120 * time.Second, maxBuffer: 32 << 20}, "gh", a...)
	}

	// Startup is retried: a transient gh hiccup at launch must not kill a
	// watch meant to run for hours.
	for attempt := 1; nameWithOwner == ""; attempt++ {
		out, err := gh("repo", "view", "--json", "nameWithOwner")
		if err == nil {
			var repo struct {
				NameWithOwner string `json:"nameWithOwner"`
			}
			if err = json.Unmarshal([]byte(out), &repo); err == nil {
				nameWithOwner = repo.NameWithOwner
				break
			}
		}
		if attempt >= 5 {
			return fail(stderr, err)
		}
		fmt.Fprintf(stderr, "warn: startup repo lookup failed (attempt %d/5); retrying\n", attempt)
		sleep(time.Duration(min(attempt*15, 60)) * time.Second)
	}
	owner, repo, _ := strings.Cut(nameWithOwner, "/")

	var prev *Snapshot
	var prevMergeable *string
	failures := 0
	for {
		out, err := gh("api", "graphql", "-f", "query="+eventsQuery, "-f", "owner="+owner, "-f", "repo="+repo, "-F", "pr="+strconv.Itoa(pr))
		var curr Snapshot
		if err == nil {
			curr, err = shapeSnapshot([]byte(out), prevMergeable)
		}
		if err != nil {
			// One failed poll must not kill the watch: log to stderr (never the
			// event stream) and back off.
			failures++
			factor := min(1<<min(failures, 3), 8)
			if rateLimitish.MatchString(err.Error()) {
				factor = 8
			}
			fmt.Fprintf(stderr, "warn: poll failed (%s); backoff x%d\n", err, factor)
			sleep(every * time.Duration(factor))
			continue
		}
		failures = 0
		prevMergeable = curr.Mergeable
		events, done := computeEvents(prev, curr)
		for _, e := range events {
			fmt.Fprintln(stdout, e)
		}
		prev = &curr
		if done {
			return 0
		}
		sleep(every)
	}
}
